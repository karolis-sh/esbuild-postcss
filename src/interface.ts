export type Options = Partial<{ extensions: string[] }>;
