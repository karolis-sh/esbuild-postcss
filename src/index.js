// eslint-disable-next-line no-var
var a = {};

export default { ...a };
