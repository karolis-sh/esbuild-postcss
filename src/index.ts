import path from 'path';
import fs from 'fs/promises';
import { Plugin } from 'esbuild';
import postcss from 'postcss';
import postcssrc from 'postcss-load-config';

import { Options } from './interface';

type PostcssOptions = Partial<postcssrc.Result>;

export = ({ extensions = ['.css'] }: Options = {}): Plugin => ({
  name: 'postcss',
  async setup(build) {
    let postcssOptions: PostcssOptions | false;

    try {
      postcssOptions = await postcssrc();
    } catch (err) {
      postcssOptions = false;
    }

    build.onLoad({ filter: new RegExp(`(${extensions.join('|')})$`) }, async (args) => {
      if (postcssOptions) {
        const css = await fs.readFile(args.path, 'utf8');

        const result = await postcss(postcssOptions.plugins).process(css, {
          ...postcssOptions.options,
          from: args.path,
        });

        return { contents: result.css, loader: 'css' };
      }

      if (path.extname(args.path) !== '.css') {
        const css = await fs.readFile(args.path, 'utf8');
        return { contents: css, loader: 'css' };
      }

      return { loader: 'css' };
    });
  },
});
