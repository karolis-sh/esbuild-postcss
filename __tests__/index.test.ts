import path from 'path';
import os from 'os';
import fs from 'fs/promises';
import { build } from 'esbuild';

import postcss from '../src';
import { Options } from '../src/interface';

const process = require('process');

const bundle = async (entry: string, options?: Options): Promise<string> => {
  const inputFilename = path.join(__dirname, entry);
  const outputDir = path.join(os.tmpdir(), 'esbuild-postcss', entry, Date.now().toString());
  const cwd = path.dirname(inputFilename);

  const spy = jest.spyOn(process, 'cwd');
  spy.mockReturnValue(cwd);
  await build({
    entryPoints: [inputFilename],
    outdir: outputDir,
    outbase: cwd,
    bundle: true,
    plugins: [postcss(options)],
    sourcemap: true,
  });
  spy.mockClear();

  return fs.readFile(
    path.join(outputDir, `${path.basename(entry, path.extname(entry))}.css`),
    'utf-8'
  );
};

it('should handle css file entry', async () => {
  const output = await bundle('fixtures/config-file/index.css');
  expect(output).toMatchSnapshot();
});

it('should handle css imports', async () => {
  const output = await bundle('fixtures/config-file/index.js');
  expect(output).toMatchSnapshot();
});

it('should handle child css import', async () => {
  const output = await bundle('fixtures/config-file/import.css');
  expect(output).toMatchSnapshot();
});

it('should handle child css import url', async () => {
  const output = await bundle('fixtures/config-file/import-url.css');
  expect(output).toMatchSnapshot();
});

it('should handle postcss-import style inlining plugin', async () => {
  const output = await bundle('fixtures/postcss-import/index.css');
  expect(output).toMatchSnapshot();
});

it('should handle postcss-import style inlining plugin via import url', async () => {
  const output = await bundle('fixtures/postcss-import/index-import-url.css');
  expect(output).toMatchSnapshot();
});

it('should handle no config file', async () => {
  const output = await bundle('fixtures/no-config-file/index.css');
  expect(output).toMatchSnapshot();
});

it('should use extensions option', async () => {
  const output = await bundle('fixtures/config-file/index.pcss', {
    extensions: ['.css', '.pcss'],
  });
  expect(output).toMatchSnapshot();
});

it('should use extensions option with no config', async () => {
  const output = await bundle('fixtures/no-config-file/index.pcss', {
    extensions: ['.css', '.pcss'],
  });
  expect(output).toMatchSnapshot();
});
