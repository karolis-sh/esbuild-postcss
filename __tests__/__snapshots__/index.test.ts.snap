// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle PostCSS options 1`] = `
"/* __tests__/fixtures/postcss-options/index.sss */
.parent > .child {
  color: black;
}
"
`;

exports[`should handle child css import 1`] = `
"/* __tests__/fixtures/config-file/_root.css */
:root {
  --color: blue;
}

/* __tests__/fixtures/config-file/features/nested.css */
.a .b {
  color: black;
}

/* __tests__/fixtures/config-file/features/properties.css */
:root {
  --background-color: yellow;
}
button {
  background-color: yellow;
  background-color: var(--background-color);
  color: var(--color);
}

/* __tests__/fixtures/config-file/import.css */
body {
  padding: 0;
  color: var(--color);
}
"
`;

exports[`should handle child css import url 1`] = `
"/* __tests__/fixtures/config-file/_root.css */
:root {
  --color: blue;
}

/* __tests__/fixtures/config-file/features/nested.css */
.a .b {
  color: black;
}

/* __tests__/fixtures/config-file/features/properties.css */
:root {
  --background-color: yellow;
}
button {
  background-color: yellow;
  background-color: var(--background-color);
  color: var(--color);
}

/* __tests__/fixtures/config-file/import-url.css */
body {
  padding: 0;
  color: var(--color);
}
"
`;

exports[`should handle css file entry 1`] = `
"/* __tests__/fixtures/config-file/index.css */
:root {
  --color: red;
}
h1 {
  color: red;
  color: var(--color);
}
.x_y {
  width: auto;
}
"
`;

exports[`should handle css imports 1`] = `
"/* __tests__/fixtures/config-file/index.css */
:root {
  --color: red;
}
h1 {
  color: red;
  color: var(--color);
}
.x_y {
  width: auto;
}
"
`;

exports[`should handle no config file 1`] = `
"/* __tests__/fixtures/no-config-file/button.css */
button {
  border: 1px solid red;
}

/* __tests__/fixtures/no-config-file/index.css */
body {
  padding: 0;
}
"
`;

exports[`should handle postcss-import style inlining plugin 1`] = `
"/* __tests__/fixtures/postcss-import/index.css */
:root {
  --color: blue;
}
:root {
  --background-color: yellow;
}
button {
  background-color: yellow;
  background-color: var(--background-color);
  color: blue;
  color: var(--color);
}
body {
  padding: 0;
  color: blue;
  color: var(--color);
}
"
`;

exports[`should handle postcss-import style inlining plugin via import url 1`] = `
"/* __tests__/fixtures/postcss-import/index-import-url.css */
:root {
  --color: blue;
}
:root {
  --background-color: yellow;
}
button {
  background-color: yellow;
  background-color: var(--background-color);
  color: blue;
  color: var(--color);
}
body {
  padding: 0;
  color: blue;
  color: var(--color);
}
"
`;

exports[`should use extensions option 1`] = `
"/* __tests__/fixtures/config-file/index.pcss */
:root {
  --color: red;
}
h1 {
  color: red;
  color: var(--color);
}
.x_y {
  width: auto;
}
"
`;

exports[`should use extensions option with no config 1`] = `
"/* __tests__/fixtures/no-config-file/button.css */
button {
  border: 1px solid red;
}

/* __tests__/fixtures/no-config-file/index.pcss */
body {
  padding: 0;
}
"
`;
